edgelet/iotedged/src/lib.rs

```diff
@@ -571,7 +582,104 @@ where
     }
 }
 
-fn set_iot_edge_env_vars<S>(settings: &S) -> Result<(), Error>
+fn retrieve_external_provisioning_info(
+    external: &External,
+    tokio_runtime: &mut tokio::runtime::Runtime,
+) -> Result<ProvisioningResult, Error> {
+    info!("Retrieving provisioning information from the external endpoint...");
+    let external_provisioning_client = ExternalProvisioningClient::new(external.endpoint())
+        .context(ErrorKind::Initialize(
+            InitializeErrorReason::ExternalProvisioningClient(
+                ExternalProvisioningErrorReason::ClientInitialization,
+            ),
+        ))?;
+    let external_provisioning = ExternalProvisioning::new(external_provisioning_client);
+
+    let provision_fut = external_provisioning
+        .provision(MemoryKeyStore::new())
+        .map_err(|err| {
+            Error::from(err.context(ErrorKind::Initialize(
+                InitializeErrorReason::ExternalProvisioningClient(
+                    ExternalProvisioningErrorReason::Provisioning,
+                ),
+            )))
+        });
+
+    tokio_runtime.block_on(provision_fut)
+}
+
+fn get_external_provisioning_info<S>(
+    settings: &S,
+    tokio_runtime: &mut tokio::runtime::Runtime,
+) -> Result<Option<ProvisioningResult>, Error>
+where
+    S: RuntimeSettings,
+{
+    if let Provisioning::External(external) = settings.provisioning() {
+        // Set the external provisioning endpoint environment variable for use by the custom HSM library.
+        env::set_var(
+            EXTERNAL_PROVISIONING_ENDPOINT_KEY,
+            external.endpoint().as_str(),
+        );
+
+        let prov_info = retrieve_external_provisioning_info(external, tokio_runtime)?;
+
+        if let Some(credentials) = prov_info.credentials() {
+            if let CredentialSource::Payload = credentials.source() {
+                if let AuthType::X509(x509) = credentials.auth_type() {
+                    let subdir_path =
+                        Path::new(&settings.homedir()).join(EDGE_EXTERNAL_PROVISIONING_SUBDIR);
+
+                    // Ignore errors from this operation because we could be recovering from a previous bad
+                    // configuration and shouldn't stall the current configuration because of that
+                    let _u = fs::remove_dir_all(&subdir_path);
+                    DirBuilder::new()
+                        .recursive(true)
+                        .create(&subdir_path)
+                        .context(ErrorKind::Initialize(
+                            InitializeErrorReason::HybridAuthDirCreate,
```

Comment:
`HybridAuthDirCreate` (and `HybridAuthKeyCreate` below) are not the right reasons for this function, since the actual hybrid key derivation happens later. These should use something from `ExternalProvisioningErrorReason`.

You can probably rename `InvalidIdentity*` to `DownloadIdentity*` and use those (base64-decoding and saving-to-file are both part of downloading).